### PR TITLE
cmd/candid: Support loading CA certificates from additional locations

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

**Comment by mhilton** [comment]
This should help resolve #128

File: candidtest/generate_certs.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
+// Copyright 2018 Canonical Ltd.
```
Comment (frankban) [inline candidtest/generate_certs.go]: 2019

File: cmd/candid/internal/admincmd/command.go [hunk 2]
```diff
@@ -144,10 +155,50 @@ func (c *candidCommand) bakeryClientLocked(ctxt *cmd.Context) (*httpbakery.Clien
...
+	for _, fn := range filepath.SplitList(os.Getenv("CANDID_CA_CERTS")) {
+		f, err := os.Open(fn)
+		if os.IsNotExist(err) || os.IsPermission(err) {
+			// Ignore the file not existing.
+			continue
+		}
+		if err != nil {
+			return errgo.Notef(err, "cannot load CA certificates")
+		}
+		defer f.Close()
```
Comment (frankban) [inline cmd/candid/internal/admincmd/command.go]: We are deferring closing all files at the end of the loop. Maybe create a closure here?

Comment (mhilton) [reply cmd/candid/internal/admincmd/command.go]: I think this is fine, If the user manages to open too many files with this mechanism then they're on their own.

File: candidtest/generate_certs.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	return params, nil
+}
+
+type genCertParams struct {
+	CACert []byte
+	Key    []byte
+	Cert   []byte
+}
+
+var genCertTemplate = template.Must(template.New("").Parse(`
```
Comment (rogpeppe) [inline candidtest/generate_certs.go]: Maybe marginally clearer to make an explicit quote function (I had to look up the ASCII for 0x60)?

        var genCertTemplate = template.Must(template.New("").Funcs(template.FuncMap{
                "quote": func(s string) string { return "`" + s + "`" },
        }).Parse(`
                var caCert = []byte({{ quote .CACert }})
        `)

File: cmd/candid/internal/admincmd/command.go [hunk 4]
```diff
@@ -144,10 +155,50 @@ func (c *candidCommand) bakeryClientLocked(ctxt *cmd.Context) (*httpbakery.Clien
...
 	return bClient, nil
 }
 
+// loadCACerts loads any certificates found in the files specified by
+// CANDID_CA_CERTS, if any, and adds them to the system CA certificates
+// for this client.
+func (c *candidCommand) loadCACerts(client *http.Client) error {
+	var certPool *x509.CertPool
+	for _, fn := range filepath.SplitList(os.Getenv("CANDID_CA_CERTS")) {
+		f, err := os.Open(fn)
```
Comment (rogpeppe) [inline cmd/candid/internal/admincmd/command.go]: how about:

     data, err := ioutil.ReadFile(fn)
     if os.IsNotExist(err) || os.IsPermission(err) {
         continue
      }

?

One thing though: why *are* we ignoring some errors here?

File: cmd/candid/internal/admincmd/command.go [hunk 5]
```diff
@@ -144,10 +155,50 @@ func (c *candidCommand) bakeryClientLocked(ctxt *cmd.Context) (*httpbakery.Clien
...
+	}
 	c.bakeryClient = bClient
 	return bClient, nil
 }
 
+// loadCACerts loads any certificates found in the files specified by
+// CANDID_CA_CERTS, if any, and adds them to the system CA certificates
+// for this client.
+func (c *candidCommand) loadCACerts(client *http.Client) error {
+	var certPool *x509.CertPool
```
Comment (rogpeppe) [inline cmd/candid/internal/admincmd/command.go]: or initialize it regardless and use `len(certPool.Subjects)` as a condition later, saving the conditional initialization in the loop, which is probably more accurate (a file might have no certs in it)

File: cmd/candid/internal/admincmd/command.go (hunk 2 above)
Comment (mhilton) [reply cmd/candid/internal/admincmd/command.go]: @rogpeppe's suggestion of using ReadFile means this is no longer a problem (not that I accept it was before)

File: cmd/candid/internal/admincmd/command_test.go [hunk 6]
```diff
@@ -87,3 +91,69 @@ func (s *fixture) Run(args ...string) (code int, stdout, stderr string) {
...
+	nonExistentFile := filepath.Join(dir, "non-existent.pem")
+
+	err = ioutil.WriteFile(certFile, srv.CACert, 0600)
+	c.Assert(err, qt.Equals, nil)
+	err = ioutil.WriteFile(emptyFile, nil, 0600)
+	c.Assert(err, qt.Equals, nil)
+	err = ioutil.WriteFile(unreadableFile, nil, 0)
+	c.Assert(err, qt.Equals, nil)
+
+	c.Setenv("CANDID_CA_CERTS", emptyFile+":"+unreadableFile+":"+nonExistentFile+"::"+certFile)
```
Comment (rogpeppe) [inline cmd/candid/internal/admincmd/command_test.go]: is the `::` deliberate here?
You've already got an non-existent file.

Comment (mhilton) [reply cmd/candid/internal/admincmd/command_test.go]: yes I wanted to see if it behaved differently with an empty file name.

**Pull request merged**